File: leetcode2/CountSortedVowelStrings/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# 16ms 100.00% 14MB 100.00%
# iterative dynamic programming
# O(N) O(1)
class Solution:
    def countVowelStrings(self, n: int) -> int:
        a, e, i, o, u = 1, 1, 1, 1, 1
        for _ in range(n-1):
            u = u
            o = o + u
            i = i + o
            e = e + i
            a = a + e
        return a + e + i + o + u

File: leetcode2/CountSortedVowelStrings/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>

//        i
//   n: 1 2 3
//        j
// sum: 3
// ans: 5

// 0ms 100.00% 6.5MB 30.00%
// iterative dynamic programming
// O(N^2) O(1)
class Solution {
 public:
  int countVowelStrings(int n) {
    int ans = 0;
    for (int i = 1; i <= n+1; ++i) {
      int sum = 0;
      for (int j = 1; j <= i; ++j) {
        sum += j;
        ans += sum;
      }
    }
    return ans;
  }
};
